### PR TITLE
PR#13 - Spring security.

### PR DESCRIPTION
**Pull request by leonplondon**
# Summary
Spring security implementation.
Added Spring Security dependencies.
Added HTTP Basic authentication.
Added test for new features.
Refactored old test that were broken because Spring Security implementation.

# Trello Card
https://trello.com/c/vIAFlkGM/23-integrar-spring-security

# Notes
PR pointing towards #12 because old PRs hasn't been merged yet.

### CODE REVIEW

File: src/main/java/wolox/training/config/SecurityConfig.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+    @Override
+    protected void configure(HttpSecurity http) throws Exception {
+        http
+            .csrf()
+            .disable()
+            .exceptionHandling()
+            .and()
+            .authorizeRequests()
+            .antMatchers(HttpMethod.POST, "/api/books/").permitAll()
```
Comment (IgnOliveto) [inline src/main/java/wolox/training/config/SecurityConfig.java]: You can apply authorization request in other way. There is a common pattern in all protected urls. Look at this [example](http://blog.florian-hopf.de/2017/08/spring-security.html)

Comment (leonplondon) [reply src/main/java/wolox/training/config/SecurityConfig.java]: Do you mean overriding the `configure(WebSecurity web) throws Exception`?

**Review by IgnOliveto — DISMISSED** [review]

**Review by rabinovichm — DISMISSED** [review]

File: src/test/java/wolox/training/TrainingApplicationTests.java [hunk 2]
```diff
@@ -1,11 +1,11 @@
...
+//
+//import org.junit.runner.RunWith;
+//import org.springframework.boot.test.context.SpringBootTest;
+//import org.springframework.test.context.junit4.SpringRunner;
+//
+//@RunWith(SpringRunner.class)
+//@SpringBootTest
+//public class TrainingApplicationTests {
+//
+//}
```
Comment (rabinovichm) [inline src/test/java/wolox/training/TrainingApplicationTests.java]: why is this all commented?

File: src/test/java/wolox/training/controllers/BookControllerTest.java [hunk 3]
```diff
@@ -0,0 +1,368 @@
...
+import wolox.training.exceptions.ParseBookException;
+import wolox.training.exceptions.RequestException;
+import wolox.training.models.Book;
+import wolox.training.models.dtos.BookDto;
+import wolox.training.repositories.BookRepository;
+import wolox.training.services.OpenLibraryService;
+
+@RunWith(SpringRunner.class)
+@WebMvcTest(BookController.class)
+public class BookControllerTest {
```
Comment (rabinovichm) [inline src/test/java/wolox/training/controllers/BookControllerTest.java:41]: Add a test checking that a non authenticated user cannot enter to an endpoint that needs the user to be authenticated.

File: src/test/java/wolox/training/controllers/UserControllerTest.java [hunk 4]
```diff
@@ -0,0 +1,434 @@
...
+import wolox.training.TestUtils;
+import wolox.training.authentication.UserAndPasswordAuthenticationProvider;
+import wolox.training.models.Book;
+import wolox.training.models.User;
+import wolox.training.repositories.BookRepository;
+import wolox.training.repositories.UserRepository;
+
+@RunWith(SpringRunner.class)
+@WebMvcTest(UserController.class)
+public class UserControllerTest {
```
Comment (rabinovichm) [inline src/test/java/wolox/training/controllers/UserControllerTest.java:43]: Add a test checking that a non authenticated user cannot enter to an endpoint that needs the user to be authenticated.

**Review by rabinovichm — APPROVED** [review]
Nice job!

**Review by IgnOliveto — APPROVED** [review]

**Pull request merged**